Spec: Add spec for expressions (#16652)
apache/iceberg: format/expressions-spec.md

 - The ASF licenses this file to You under the Apache License, Version 2.0
 - (the "License"); you may not use this file except in compliance with
 - the License.  You may obtain a copy of the License at
 -
 -   http://www.apache.org/licenses/LICENSE-2.0
 -
 - Unless required by applicable law or agreed to in writing, software
 - distributed under the License is distributed on an "AS IS" BASIS,
 - WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 - See the License for the specific language governing permissions and
 - limitations under the License.
 -->

# Iceberg Expressions

This document defines the structure and behavior of expressions for use in Iceberg specifications. The purpose is to define a common structure that enables simple expressions to be stored and exchanged.

Stored expressions are needed for use cases like data validations (`CHECK` constraints) and default values (for instance, `current_timestamp()`). Expressions are exchanged in use cases like server-side scan planning in the catalog protocol.

## Overview

The goal of this specification is to define a simple expression structure and avoid complexity.

> To remain simple, the expressions that can be represented are deliberately constrained. Value expressions are constants, field references, or function calls with value expression arguments. Predicates are comparisons of value expressions that produce true or false.

Reviewer: ```suggestion
To remain simple, the expressions that can be represented are deliberately constrained.
- Value expressions:  constants, field references, or function calls with value expression arguments. 
- Predicates: comparisons of value expressions that produce true or false.
```

I think bulleting here would help emphasize the two types/categories

Reviewer: (looks like you did this just below, so maybe we can simplify this to just the value expressions and predicates and let the definition below stand-in for the more complete version).

Author: Since I have a specific section below, I made this a shorter overview rather than duplicating the more detailed version.

> This approach is intended to keep focus on the logical structure of expressions. Complexity is pushed to the functions that are called, which can be a limited set of well-defined and portable functions (like Iceberg partition transforms) or could be user-defined functions that can use the full range of SQL capabilities. Multi-dialect UDFs are responsible for any SQL constructs that are specific to an engine, rather than importing and duplicating dialects in Iceberg expressions.

Reviewer: ```suggestion
This approach is intended to keep focus on the logical structure of expressions. Complexity is pushed to the functions that are called, which are be a limited set of well-defined and portable functions (like Iceberg partition transforms) or user-defined functions that can use the full range of SQL capabilities. Multi-dialect UDFs are responsible for any SQL constructs that are specific to an engine, rather than importing and duplicating dialects in Iceberg expressions.
```
Suggestion: make the language a little less uncertain

If we have a UDF reference this might be a good place to include it.

Author: Updated.

> This is consistent with Iceberg's conservative approach in other specs. Expressions and predicates are an important part of Iceberg implementation APIs, but have been deliberately limited in specifications. For example, sort orders and partition fields are strictly limited to a small set of transforms over well-defined inputs (source field IDs). This spec is widening what can be expressed, but depends on function calls for complex tasks.

Reviewer 2: More of a "why this is written in the way it is" than a "what this spec is about". Just wondering if we need this paragraph in the text.

Author: That's fair. I can remove it if you think it isn't useful.

Reviewer 2: My goal would be the Spec is just what would be needed to implement expressions correctly. So I would remove or maybe just start a whole new doc about Iceberg's core values.

Reviewer 3: Yeah not strongly opinionated here but I find this to just be extra stuff that isn't super valuable in the context of the spec itself. A separate core values or design principles would be valuable though.

This specification covers the structure of Iceberg expressions and includes appendicies that specify serialization as JSON and a set of portable functions defined by Iceberg specifications.

## Structure

Iceberg expressions have two types:

* **Value expressions** represent data values and transformations of values (function calls) that produce any Iceberg type

> * **Predicates** represent comparisons of value expressions and boolean logic that produce `true` or `false`

Reviewer 4: This phrase is ambiguous: "comparisons of value expressions and boolean logic" can be misread as "comparisons of (value expressions and boolean logic)" — i.e., comparing value expressions against boolean logic. Suggest splitting:

> **Predicates** are formed from comparisons of value expressions, combined with boolean logic, and produce `true` or `false`.

Author: Updated to:
> **Predicates** represent comparisons of value expressions as well as combinations of predicates with boolean logic (and, or, not)

### Value expressions

> A value expression is an expression that produces a typed value

Reviewer: ```suggestion
A value expression is an expression that produces a typed value.
```

Author: Fixed.

Value expressions can be one of three types: a constant value, a field reference, or a function applied to zero or more value expressions.

#### Constant values

A constant or literal is the simplest type of value expression that represents a specific typed value.

> #### Field reference

Reviewer: In looking around, other systems have two other reference types that I want to callout (though I don't necessarily think we want to include, but should consider):
1. Positional References (for row-like references)
2. Subscripted References (indexing into arrays)

Author: For subscripted references, we can use a well-defined function, like `get_item`. I don't think that we want to have this in the spec since it fits cleanly into a function. And that would also requiring choosing either 1-based or 0-based indexing. With functions, we can model both: `get_item_the_right_way` for 0-based, `get_item_sql` for 1-based. 😄 

I don't think that we want to include positional references.

Author: Looks like Trino uses `element_at`: https://trino.io/docs/current/functions/array.html#element_at

A field reference represents the value of a specific field in a row. When an expression is evaluated on a row, it returns the value of the field.

Field references may be named references (unbound) or ID references (bound). ID references identify a field by field ID from a schema. Named references identify a field by name that must be resolved to an ID (bound to a schema) to access the field.

> ID references are used for stored expressions, where the identity of the column is determined when the stored expression is created. For example, column constraints are tied to field ID so that renaming a column does not drop its stored constraint.

Reviewer: ```suggestion
ID references are used for stored expressions, where the identity of the column is determined when the stored expression is created. For example, column constraints are tied to field ID so that renaming a column does not incorrectly reference its stored constraint.
```

Not sure this is the right wording, but 'drop' feels confusing in this context.

Author: I think it should probably be "invalidate the reference in"

Reviewer 2: or "invalidate the reference to its stored constraint"

Author: Fixed.

> ID references are used for stored expressions, where the identity of the column is determined when the stored expression is created. For example, column constraints are tied to field ID so that renaming a column does not drop its stored constraint.

Reviewer 2: ```suggestion
ID references are used for stored expressions, where the identity of the column is determined when the stored expression is created. For example, column constraints are tied to field IDs so that renaming a column does not drop its stored constraint.
```

Author: Fixed.

Named references are used when the identity of the column is determined when the expression is evaluated. For example, query filters are resolved each time a query runs so servers-side planning uses unbound named references.

The context in which an expression is used determines the type of references that are valid. Iceberg specifications should document whether ID references, named references, or both are allowed.

#### Apply function

An apply expression represents the result of a function applied to (or called on) zero or more values produced by child value expressions.

Functions are identified by catalog, namespace, and name.

* Function name is always required

> * Namespace is optional and is assumed to be empty ([]) if it is not present or is null

Reviewer 2: What does the parenthetical for empty mean here? Are we defining [] as an empty set?

Author: An empty list, since a namespace is a list of identifier parts.

Is there an edit you would suggest to be more clear?

Author: I updated this to be consistent with catalog object identifiers, which removed this issue. Now, the statement is that functions are referenced by a catalog and a multi-part identifier with at least one part, the function name. That avoids needing to make this rule.

* Catalog is optional and is assumed to be the catalog in which the referencing object is stored if it is not present or is null

> The catalog name is used to identify the catalog where the function definition can be loaded or it identifies a reserved function set. As in the view and UDF specs, catalog names represent connection configurations that may differ across environments. Omitting catalog names is recommended to avoid depending on consistent environments. For example, if a table has a CHECK constraint that references a UDF without a catalog name (missing or null), the UDF should be loaded from the table’s catalog.

Reviewer 2: I'm a little confused about this paragraph. We have the concept of a Catalog and Function Set. It feels like a "catalog" is a superset of the concept of "function set" but this feels like it is hard to disambiguate with a "namespace" which is also a "function set"

Maybe if on line 86 it just said "reserved catalog names are"

Author: I was trying to avoid calling reserved function sets "catalogs". Maybe I could change it to "Reserved function sets are identified by fixed catalog names:"?

Author: Updated to:
> The catalog name identifies the catalog where the function definition can be loaded or is a reserved name that identifies a set of functions.

And:
> The reserved names used to identify sets are:

Reserved function set names are:

* `sql_functions` is used for functions defined by the SQL standard
* `iceberg_functions` is used for functions defined in this specification

Engines may document and use a catalog name to identify their built-in functions that are not part of the SQL spec, like `spark_builtin_functions.to_utc_timestamp`.

> Producers are responsible for resolving catalog, namespace, and name if the environment is relevant. For example, if a SQL engine uses its current catalog and namespace to find a function, the resolved catalog and namespace must be used to produce an unambiguous function identifier.

Reviewer 2: I'm not sure I understand this statement either. Is this just saying that a engine is allowed to resolve an identifier however it likes as long as it would be doing so unambigously?

Author: Clarified by adding:
> Function identifiers are unambiguous and are not interpreted using session context.

What I am trying to say is that identifiers are fully-qualified and identify a single object. In a SQL query, I could use a function like `parse_device_id`, which is resolved using my session's current catalog and namespace to `prod.website.parse_device_id`. The identifier when referencing that function must be `{"catalog": "prod", "namespace": ["website"], "name": "parse_device_id"}`. The identifier is not dependent on session context.

I should also note that `"catalog"` could be omitted to use the catalog where the referencing object is stored.

#### Value expression types

The type produced by a value expression may change. For example, an ID reference may produce a widened type after the underlying column's type is promoted.

> Function calls may produce different types when function definitions change, and type changes may change the definition that is resolved for a function name. For example, `identity(int) -> int` will change to `identity(long) -> long` when an input field is promoted from `int` to `long`.

Reviewer: I'm a little confused on the wording here.  Are we saying that functions will automatically widen (I don't think that's the intent, but it feels like that's what this is saying).

Isn't it more that we're parameter matching when binding the function based on input types and the corresponding return type can change?

Reviewer: next line seems to indicate that I got the intent right, but the wording feels a little confusing in terms of how we're describing it.

Author: I'll update this to state that the identity function that gets resolved is different.

> A value expression's type is determined when it is bound to a specific input schema.

Reviewer: ```suggestion
A value expression's return type is determined when it is bound to a specific input schema.
```

Author: Note to myself: I also want to note that there are some cases where you want to track an expected return type. For example, expressions that produce stats need an expected type to produce a `content_stats` schema.

Author: I updated this to "result type" rather than "return".

> If types are incompatible at runtime, implementations binding or evaluating expressions may apply type promotion to align types for predicates and to resolve functions. Implementations may choose when to promote values to accomodate engines that differ in casting behavior. However, implementations must fail rather than insert "unsafe" casts. 

Reviewer: Is this always implicitly handled by the engine? Do we consider `CAST(...)` something that would fall under a `sql_functions.cast` or should we leave it out entirely.

Author: I think that `sql_functions.cast` can be used for explicit casts, but I want to avoid adding a cast definition in this spec for the expression structure.

> If types are incompatible at runtime, implementations binding or evaluating expressions may apply type promotion to align types for predicates and to resolve functions. Implementations may choose when to promote values to accomodate engines that differ in casting behavior. However, implementations must fail rather than insert "unsafe" casts. 

Reviewer 2: I understand the intent here but adding "unsafe" feels ambiguous here.

Author: We used unsafe in the UDF spec as well:
> Require explicit casts for unsafe or non-obvious conversions.

I'll remove the quotes because that makes this more confusing. We do delegate to engines what casts they want to produce, but I think it's fairly well understood in most cases.

### Predicates

A predicate is a boolean expression that produces true or false.

Predicates can be constants (true or false), comparisons or tests of value expressions, or logical combinations of predicates (AND, OR, NOT).

> If value expression types in a predicate are incompatible, implementations should align types using type promotion. For instance, `int_col > 5.0` should promote int values to float. If the types cannot be aligned according to type promotion rules, the predicate must evaluate to false. For instance, `"goats" > -Infinity` should always be `false`.

Reviewer 2: I feel like we are again dancing around what it is a "safe" type promotion and what is allowed. The above examples make sense to me but for me the dangerous ones are always things like

`2141041920` > '05-04-19` or whatnot

Author: What I'm trying to say is that implementations should fail the dangerous cases.

I get the argument that this is not exact, but we are delegating this responsibility to engines so that their rules are applied consistently. I think that most engines have well-defined lines between safe and unsafe casts. But there are some situations like promoting decimals to wider scale that could differ.

> If value expression types in a predicate are incompatible, implementations should align types using type promotion. For instance, `int_col > 5.0` should promote int values to float. If the types cannot be aligned according to type promotion rules, the predicate must evaluate to false. For instance, `"goats" > -Infinity` should always be `false`.

Reviewer 4: Two issues with this paragraph:

**1. "type promotion rules" is undefined here.** The Iceberg type promotion rules live in the table spec; please add a cross-reference so readers know which exact rules govern alignment.

**2. Conflict with L104.** This paragraph says incompatible types `must evaluate to false`, but the [Value expression types](#value-expression-types) section four lines up says implementations `must fail rather than insert "unsafe" casts`. Both rules apply to the example given here.

Take `"goats" > -Infinity`: string and float have no promotion path — there is no common type to align them to.
- Following L113, this predicate must return `false`.
- Following L104, the implementation must fail because no safe cast exists.

Same expression, two contradictory required behaviors, four lines apart. Two engines on the same data and predicate will diverge — one throws, one returns zero rows — silently.

The spec probably has two distinct cases in mind that should be split:

- **No promotion path exists at all** (e.g., `string` vs `float`) → short-circuit the predicate to `false`. Comparisons across unrelated type families just do not match.
- **A promotion path exists but the cast would be lossy/unsafe** (e.g., `long` → `int` truncation, high-precision `decimal` → `float`) → fail rather than silently produce wrong results.

That is a coherent model, but it is not in the text. Either pick one rule and apply it uniformly to type-incompatible expressions, or distinguish the two cases explicitly.

Author: The promotion rules for value expressions are delegated to engines, as is stated in the previous section. That's why type promotion rules are not defined here.

For the second point, the failure required by line 104 was intended for resolution instead of predicate evaluation. I'll add an example, like `sqrt(double)` called as `sqrt("orange")` would fail because the function cannot be resolved.

That said, I'm now rethinking the requirement for predicates. I thought initially that incomparable types would clearly result in `false` (like `"goats" > -Infinity`) but that's not what the reference implementation does. More importantly, it breaks reasonable relationships since `"goats" > -Infinity` is `false` would normally imply that the negation (`-Infinity >= "goats"`) is `true`. And it's the same for `!=`, where by the logic of making the incompatible type comparison false, these values are clearly not equal and this should be true.

I think I'm going to change this so that incomparable types should fail. If there are objections, please comment.

Reviewer 4: Fail seems to the correct call. I agree with the two above arguments on failing incompatible types in predicate: (1) The negation-symmetry break is the decisive one (2) match the current reference implementation rather than introducing a new behavior.

Two additional supporting arguments:

**Many analytic engines fail on type-incompatible comparisons** :

- Spark SQL (ANSI mode): `Incompatible types` at analysis when no common type is reachable in the [precedence list](https://spark.apache.org/docs/latest/sql-ref-ansi-compliance.html).
- Trino: planner-time error — the [comparison docs](https://trino.io/docs/current/functions/comparison.html) make this explicit on `BETWEEN`: *"Trino produces an error if you ask it if `John` is between `2.3` and `35.2`."*
- Snowflake: *"When conversion isn’t possible, implicit conversion fails"* — [Data Type Conversion](https://docs.snowflake.com/en/sql-reference/data-type-conversion), illustrated by `COALESCE('foo', 1)` erroring because `'foo'` cannot convert to NUMBER.

**Type-incompatibility at evaluation is a binding bug, not a row condition.** Once an expression is bound to a schema, incompatibility means the expression was malformed at binding, or producer and consumer disagree on a column's type. Silent false turns that structural bug into a zero-row result; failing surfaces it.

> Value expressions are not valid predicates, even when the expression is expected to return a boolean value. Value expressions must be compared or tested to produce a predicate. For example, `is_empty("")` is not a valid predicate, but `is_empty("") = true` is a valid predicate.

Reviewer: I feel like we want to say the distinction is that predicates are two-value boolean logic and a value expression that returns boolean is three-value boolean logic, which is why you can't use value expressions for predicates.  Right?

Author: Mainly, yes. We also don't know that the function will always produce a boolean since types are determined after functions are resolved, so we want to have the type alignment rules to fall back on.

Reviewer 2: This also threw me a bit, so I would appreciate a note in the spec.

Author: Updated to state that `is_empty(str_col)` is not valid because it may produce null.

#### Comparisons

Comparisons are predicates that compare two value expressions with the same primitive type. Comparisons are:

| Comparison  | Description |
|-------------|-------------|
| `=`         | Is equal |
| `!=`        | Is not equal |
| `<`         | Less than |
| `<=`        | Less than or equal |
| `>`         | Greater than |
| `>=`        | Greater than or equal |

> Primitive types are compared using natural order, except for the following types:

Reviewer: Do you have a reference for "natural order".  This is somewhat confusing outside the context of Java since some definitions of natural comparison would be: `"i2" > "i10" -> false`.

Author: Yeah, I was wondering how to put this. Looks like Parquet uses "signed comparison" ([thrift](https://github.com/apache/parquet-format/blob/master/src/main/thrift/parquet.thrift#L1081)), I'll update to use that instead.

* `false` is less than `true` for `boolean`
* `fixed` and `binary` use unsigned byte-wise comparison

> * `string` uses unsigned byte-wise comparison of the UTF-8 representation

Reviewer 4: This rule is not derived from an IEEE or RFC standard. Should we add a short clarifying note so readers do not look for a normative reference? e.g.:

> `string` uses unsigned byte-wise comparison of the UTF-8 representation. This preserves Unicode code-point order and is independent of locale; it is **not** the Unicode Collation Algorithm ([UTS#10](https://www.unicode.org/reports/tr10/)).

Author: Added "it is not the Unicode Collation Algorithm".

Collations would be handled by comparison functions, like `compare_unicode(a, b, "collation-sequence-id")`.

* `uuid` uses unsigned byte-wise comparison of the UUID bytes

> * `float` and `double` use IEEE 754 total order after normalizing NaN to the canonical NaN (sign bit 0, exponent bits all 1, matissa msb 1 followed by all 0)

Reviewer 2: Just wondering why we don't want to follow the IEE standard here and do -Nan < Nan?

Author: We use NaN counts in stats because IEEE total ordering can easily destroy the effectiveness of lower/upper bounds with just one NaN value. Since we treat all NaN values as the same thing in our stats, I wanted to be consistent here and also push people to use `IS NAN` and `NOT NAN` predicates instead of `= NaN`.

I think we have to support NaN here; the non-total-order IEEE comparison causes a ton of headaches because NaN != NaN. But I wanted to keep the intrusion minimal. Plus, "negative not a number" is a thing because of the bit representation. And I don't think we want to keep signaling NaN here either, which would be a whole class of cases where apparent NaN < NaN is true or false, or just confusing because you don't display the mantissa.

> * `float` and `double` use IEEE 754 total order after normalizing NaN to the canonical NaN (sign bit 0, exponent bits all 1, matissa msb 1 followed by all 0)

Reviewer 4: Two issues: a typo and a framing one.

**Typo:** `matissa` → `mantissa`.

**Framing.** : is the below version more clear?

> `float` and `double` are compared by treating all NaN bit patterns as equal to each other and greater than every non-NaN value. Implementations may achieve this by canonicalizing NaN bit patterns to a single value (sign bit 0, exponent bits all 1, mantissa msb 1 followed by all 0) before applying IEEE 754 total order.

Author: Fixed the typo.

I don't think that wording is more clear. It says the same thing in the second sentence, but adds confusion by talking about treating distinct values of NaN the same, but also normalizing them.

    * `NaN = NaN` is true for any two NaN values
    * `val < NaN` is true for all non-NaN values

Note type alignment produces `decimal` values with the same scale so that comparison is equivalent to the natural order of the unscaled numeric value.

Tests are predicates that test a single value expression, optionally using a constant or set of constants. Constants must have the same type and must be non-null. Tests are:

| Test                    | Allowed types | Constant type | Description |
|-------------------------|---------------|---------------|-------------|
| `IS NULL`               | any           |               | true iff the value is null |

> | `IS NOT NULL`           | any           |               | true iff the value is not null |

Reviewer 5: minor: may be we should link supported iceberg data tyes in the spec

| `IS NaN`                | float, double |               | true iff the value is an IEEE 754 NaN |
| `IS NOT NaN`            | float, double |               | true iff the value is not an IEEE 754 NaN |
| `STARTS WITH const`     | string        | string        | true iff the constant is a prefix of the value |
| `NOT STARTS WITH const` | string        | string        | true iff the constant is not a prefix of the value |
| `IN (constant set)`     | any           | same as value | true iff the value is equal to any constant |

> | `NOT IN (constant set)` | any           | same as value | true iff the value is not equal to all constants |

Reviewer 4: Two coverage gaps in the tests table:

1. **Allowed types `any` for `IN`/`NOT IN`** — but the comparison rules above restrict equality to primitives, and the equality semantics for complex types (struct/list/map) are not defined anywhere in this spec. Either narrow the allowed types to primitive (matching `=`), or define how element equality works for non-primitive types.

2. **Null on the value side for tests is unspecified.** The null-safe rules below cover comparisons (`=`, `<`, etc.) but not tests. What does `null IS NaN` evaluate to? `null IS NOT NaN`? `null STARTS WITH 'foo'`? `null IN (1, 2, 3)`? Reading the rows literally ("true iff value is not NaN"), `null IS NOT NaN` is true, which contradicts the null-safe spirit elsewhere. A row in the null-safe section stating the rule for tests on null values — e.g., "tests on a null value evaluate to false except for `IS NULL` (true) and `IS NOT NULL` (false)" — would close this.

Author: Updated to "any primitive".

For the second, I don't think this is confusing. A null value is not equal to a NaN value. IS NOT NaN is true.

#### Boolean logic

Predicates must use 2-valued boolean logic. Evaluation of all predicates must produce `true` or `false`.

> Engines that implement SQL 3-valued boolean logic must add `IS NULL` and `NOT NULL` to produce the 2-valued equivalent. This avoids bugs in engines and languages that do not natively implement 3-valued logic. For example, the SQL predicate `x < 10` should be passed as `x < 10 AND x IS NOT NULL` for a SQL `WHERE` condition (or `x < 10`; see null-safe comparisons below). For a `CHECK` constraint, the expression is passed as `x < 10 OR x IS NULL`. This ensures that implementations will make the correct determination, rather than depending depending on context to interpret a null result (`WHERE` vs `CHECK`).

Reviewer 4: typo: `depending depending`.

Logical combinations are boolean operators applied to predicates. `AND` and `OR` are binary operations and `NOT` is a unary operation.

> Comparisons must be null-safe. For example:

Reviewer 4: The bullets below are exhaustive for `=`/`!=` but only partial for ordering operators (e.g. `null < 34`, `null <= 34`, `null > 34`, `null >= 34`, `34 > null`, `34 >= null`, `null > null`, `null >= null` are missing). 

Suggest replacing the bullets with explicit rules, then keeping a few illustrative examples:

> Comparisons must be null-safe. For any two operands `a` and `b`:
> - `a = b` is true if both are null, or both are non-null and equal; otherwise false.
> - `a != b` is the boolean negation of `a = b`.
> - `a < b` and `a > b` are false whenever either operand is null; otherwise they use the natural order defined above.
> - `a <= b` is `(a = b) OR (a < b)`; `a >= b` is `(a = b) OR (a > b)`. Both are true when both operands are null and false when exactly one operand is null.

Examples (now derivable from the rules):
> - `null = null` → true; `34 = null` → false
> - `null != null` → false; `34 != null` → true
> - `null < null` → false; `34 < null` → false
> - `null <= null` → true; `34 <= null` → false; `null <= 34` → false

Author: These are examples. Isn't it sufficient to say that comparison must be null-safe?

Author: I added these rules for clarity.

* `null = null` is `true`
* `34 = null` is `false`
* `null != null` is `false`
* `34 != null` is `true`
* `null < null` is `false`
* `null <= null` is `true`
* `34 < null` is `false`

> Comparisons must handle null values when value expressions evaluate to null. However, value expressions used to define a predicate should not directly contain null constants and may reject them. For example, `x = get_item(map, "key")` is valid although `get_item` may return a null value, but `x = null` should be rejected because `x IS NULL` is the recommended unambiguous predicate.

Reviewer: >However, value expressions used to define a predicate

Context here is a little confusing since we say at line 115 we say: Value expressions are not valid predicates.

Author: This is referring to the value expressions on the left or right side of a predicate. I'll update it.

> Comparisons must handle null values when value expressions evaluate to null. However, value expressions used to define a predicate should not directly contain null constants and may reject them. For example, `x = get_item(map, "key")` is valid although `get_item` may return a null value, but `x = null` should be rejected because `x IS NULL` is the recommended unambiguous predicate.

Reviewer 2: Wondering about the "should" paragraph here which advises folks to use "IS NULL" vs "= null" why not just make that a "must"?

Author: I just thought it would be too difficult to state clearly and concisely. You'd have to say that a value expression directly referenced in either side of a comparison may contain or produce null, but cannot be a null constant.

Explaining that went alright so I'll update this.

### Compatibility with REST catalog expressions

Older clients use more restrictive forms of predicates and references that used a "term" for specific transforms and named references. These expressions should be supported for backward compatibility to allow older clients to interact with newer REST catalog services.

> Prior to this spec, deprecated expressions were passed in the REST API in 3 places:

Reviewer 4: "deprecated expressions" is awkward here for two reasons:

1. **Anachronism.** "Prior to this spec, deprecated expressions were passed…" — these forms were not deprecated prior to this spec; they were the only form. They are being deprecated *by* this spec.

2. **Forward reference without a hook.** A reader hitting "deprecated expressions" at this point has no idea what is being deprecated until they reach `DEPRECATED_PREDICATE` / `DEPRECATED_REF` in [Appendix B](#appendix-b-json-serialization). The form should be named here so the section stands on its own.

Suggested rewrite:

> Prior to this spec, REST APIs used a more restrictive, term-based form of predicates and references in three places. Those forms are now deprecated (see [Backward compatibility](#backward-compatibility) in Appendix B):

This also lets you drop the redundant first paragraph of the section, since the term-based form is now named directly.

Author: Updated to a form of this.

* As `filter` passed to server-side scan planning
* As `filter` passed to the service in `ScanReport`
* As `residual` passed to the client with a scan task

Both server-side scan planning and the report endpoint can continue to accept filters from older clients without issues by parsing term-based expressions (see [Appendix B: JSON serialization](#appendix-b-json-serialization)).

> Residuals passed from services back to clients that do not use the new syntax would cause clients to fail, but services are allowed to omit the residual so that it is calculated on the client side (intended to avoid duplicating large IN filters). For compatibility, REST services should detect client versions and produce deprecated predicates, or omit residuals from tasks.

Reviewer 5: > REST services should detect client versions and produce deprecated predicates

unfortunately there is no reliable way to do this, but the scenario here is might be a bit narrow for example we know client is older when they send older expressions here (which referenced things by name ?) and it can produce the output. for example scan planning we expect the filter ... and the metric report is one ways where client is just trying to persist the report to the server ? 

Do we wanna a bit specific ?

Reviewer: Another case where we may want to version the endpoint to explicitly break which form of expression we're using.

Author: If you can't tell whether a client is safe, then omit the residual since it is allowed. We could say that directly, but I don't want to require it if you know the client is safe.

Author: I've updated the wording. It now says you should omit the predicate, but may send it if you can verify the client (for example, using version number). That's more clear.

## Appendix A: Iceberg functions

This section defines the functions in the `iceberg_functions` reserved catalog name.

> * `if_else(condition: predicate, when_true: T, when_false: T) -> T`: returns the value of `when_true` when `condition` is true and `when_false` otherwise

Reviewer 5: are we calling `if_else` as function ? 
- can they be nested ? if_else(condition_1, if_else(condition_2, when_true, when_false, when_true_outer) ?  return when_true when condition_1 && condition_2 is true ? 

+1 on keeping the data types same !

Reviewer: Yes, I think this how you would model `CASE WHEN` statements

Reviewer 6: nit: Shall we explicitly state `if_else` is lazy like `CASE WHEN` (only the selected branch is evaluated)? As written it's a normal apply function, implying strict evaluation of both branches, which would break guards like `if_else(d = 0, 0, divide(n, d))`.

> ### Partition transforms

Reviewer 5: [doubt] thoughts of extracting this to a seperate functions page we can add actions there too...  i know we did discuss functions.md too wondering if we wanna do this here or we can always do separately

Author: I think it makes sense to have them here. This spec isn't too long so I'd just keep them co-located so they are easy to find.

Iceberg partition transforms are also defined as functions (other than `void`).

All partition transforms produce `null` for a `null` input value.

> 

Reviewer 2: nit: "unparameterized partition transforms"? I just had a double take when I went through this table without seeing the one below

Author: This is a separate paragraph that applies to all partition transforms. Maybe it should be moved to be more clear?

|-------------------|--------------------------------------------------------------|----------------------------------------------------------------------|-------------|
| `identity(value)` | Source value, unmodified                                     | Any primitive except for `geometry`, `geography`, and `variant`      | Source type |
| `year(value)`     | Extract a date or timestamp year, as years from 1970         | `date`, `timestamp`, `timestamptz`, `timestamp_ns`, `timestamptz_ns` | `int`       |
| `month(value)`    | Extract a date or timestamp month, as months from 1970-01-01 | `date`, `timestamp`, `timestamptz`, `timestamp_ns`, `timestamptz_ns` | `int`       |
| `day(value)`      | Extract a date or timestamp day, as days from 1970-01-01     | `date`, `timestamp`, `timestamptz`, `timestamp_ns`, `timestamptz_ns` | `date`      |
| `hour(value)`     | Extract a timestamp hour, as hours from 1970-01-01 00:00:00  | `timestamp`, `timestamptz`, `timestamp_ns`, `timestamptz_ns`         | `int`       |

Note that `year`, `month`, and `hour` transforms produce ordinal values and not human-readable values. For example, `year(2018-05-13)` produces `48`, not `2018`.

> Parameterized functions are called as 2-argument functions. The first argument is an `int` parameter (`N` or `W` from the table spec) and the second argument is the value to transform. For example, `bucket(256, id)` calls `bucket[256]`.

Reviewer: The way we reference 'Parameterized functions' is a little awkward since a parameterized function is a function with one or more parameters.  So it feels strange to read as these are 2-arg functions?

Wording just makes it feel wrong.

Author: I'll just change it to bucket and truncate, since those are the only two.

| Parameterized function name | Description                                   | Source types                                                                                 | Result type |
|-----------------------------|-----------------------------------------------|----------------------------------------------------------------------------------------------|-------------|

> | `bucket(N, value)`          | Hash of value, mod `N` (see table spec)       | Any primitive except for `geometry`, `geography`, `variant`, `boolean`, `float`, or `double` | `int`       |

Reviewer: Do cross doc references work with how we publish?  Something like `[Link Text](relative/path/to/file.md#heading-anchor)` is supposed to work in markdown.  It would be nice to have pointers.

Author: I don't know how to do the refs, but I can add them.

| `truncate(W, value)`        | Value truncated to width `W` (see table spec) | `int`, `long`, `decimal`, `string`, `binary`                                                 | Source type |

## Appendix B: JSON serialization

Iceberg expressions are serialized as JSON objects in table, view, and UDF metadata, and in the REST protocol for catalogs.

> ### Value expressions

Reviewer: It would be nice to have some actual json examples. 

Any specifics on how we handle types in json?  For example, Json has no concept of int vs long value.  The specific type is erased and Json has some weird behavior around the 53-bit boundary for numeric values.  I believe they even recommend storing as a string for large values to avoid these issues.

Author: I'll improve linking, but this relies on the single-value JSON serialization in the Iceberg table spec (used for default values). That solves the serialization issues as long as you have a type.

This also adds an optional `"data-type"` field to specify the primitive type. That's optional because there are a lot of cases where you don't necessarily know the concrete type and want type alignment in binding to handle it. For instance, if a user passes `Expressions.lt("x", 5)` then it's reasonable to pass `{"type": "literal", "value": 5}` because the actual type hasn't been determined by binding.

```
EXPR: LITERAL | REFERENCE | APPLY

LITERAL: VALUE
    | { "type": "literal", "value": VALUE }
    | { "type": "literal", "value": VALUE, "data-type": DATA_TYPE }
LITERALS: [ LITERAL* ]

REFERENCE: BOUND_REF | UNBOUND_REF
BOUND_REF: ID | { "type": "reference", "id": ID }
UNBOUND_REF: NAME | { "type": "reference", "name": NAME }

> APPLY: { "type": "apply", "func-name": FUNC_ID, "arguments": [ EXPR* ] }

Reviewer 4: The grammar and Appendix A disagree on whether `if_else` can be encoded.

The grammar restricts `APPLY` arguments to value expressions: `APPLY: { "type": "apply", "func-name": FUNC_ID, "arguments": [ EXPR* ] }`, where `EXPR` is `LITERAL | REFERENCE | APPLY`. Predicates are explicitly *not* `EXPR` (see L115: "Value expressions are not valid predicates, even when the expression is expected to return a boolean value").

But Appendix A defines `if_else(condition: predicate, when_true: T, when_false: T) -> T`, with the first argument typed as a predicate.

Concrete: how do you serialize `if_else(x > 5, "big", "small")`?

```json
{
  "type": "apply",
  "func-name": "if_else",
  "arguments": [
    { "type": "gt", "left": "x", "right": 5 },
    "big",
    "small"
  ]
}
```

The first argument is a `CMP_OP` predicate, but `APPLY.arguments` requires every element to be `EXPR`. The grammar has no valid encoding for `if_else`’s condition.

Two viable fixes:

1. **Broaden `APPLY` arguments to allow `PREDICATE`** — `arguments: [ (EXPR | PREDICATE)* ]`, with function signatures declaring which positions accept which. Cost: muddies the value/predicate split that L115 enforces.
2. **Give `if_else` a dedicated JSON form** rather than treating it as an `apply` — e.g. `{ "type": "if-else", "condition": PREDICATE, "when-true": EXPR, "when-false": EXPR }`. Cost: `if_else` is no longer "just a function"; the iceberg_functions table needs a special case.

Either is fine, but the grammar and the function table need to agree — the spec’s own named example function is currently unrepresentable under its JSON grammar.

Author: I think you're right that we want `APPLY` to allow predicates as arguments.

I don't want to add `PREDICATE` to `EXPR` because that would allow predicates with left or right operands that are predicates directly, like `5 < x < 7`, which is confusing because it is a boolean compared with an integer and also creates another case where null handling has to be specified.

Author: I've introduced `FUNC_ARG: EXPR | PREDICATE` to fix this.

FUNC_ID: NAME

>     | { "catalog": NAME, "namespace": [ NAME* ], "name": NAME }

Reviewer 4: With the flat list [CatalogObjectIdentifier PR](https://github.com/apache/iceberg/pull/16144) merged, we probably should update this to a flat list too. 

[Functions spec PR](https://github.com/apache/iceberg/pull/15180) is also updated.

Author: I updated this to `"identifier": [ NAME* ]` and update the text above. Please have a look to make sure the updates make sense.

ID: integer
NAME: string

VALUE: non-null single value JSON from the table spec
DATA_TYPE: Iceberg type from the spec

> ```

Reviewer 4: The bare-value forms create a real ambiguity. `LITERAL` admits a bare `VALUE`, `BOUND_REF` admits a bare `ID` (integer), and `UNBOUND_REF` admits a bare `NAME` (string). The productions overlap exactly: a bare integer in `EXPR` position is both a valid int literal and a valid bound reference; a bare string is both a valid string literal and a valid unbound reference.

Concrete example:

```json
{ "type": "eq", "left": "name", "right": "Alice" }
```

A conformant decoder has no way to tell whether `"name"` is a reference to the field named `name` or the literal string `"name"` — and the same for `"Alice"`. Two implementations following the grammar literally could disagree on what this document means. Same problem for `{"type": "eq", "left": 42, "right": 5}`.

We can drop the bare forms for `BOUND_REF` / `UNBOUND_REF` and require `{"type": "reference", ...}`. Bare integer/string is then always a literal. Unambiguous; cost is verbosity for references, which is likely the more common case.

Author: I agree. I'm dropping the bare forms from `BOUND_REF` and `UNBOUND_REF`.

I initially left these in because the deprecated `TERM` structure supported them. But that was more restrictive and we knew that `TERM` could not be a literal. We'll have to keep it there for compatibility, but removing it from the new reference rules is a good idea.

Value expressions are not valid predicates, even when the expression is expected to return a boolean value. Value expressions must be compared or tested to produce a predicate. For example, `is_empty(str_col)` is not a valid predicate because it may produce `null`, but `is_empty(str_col) = true` is a valid predicate.

#### Comparisons

Comparisons are predicates that compare two value expressions with the same primitive type. Comparisons are:

| Comparison  | Description |
|-------------|-------------|
| `=`         | Is equal |
| `!=`        | Is not equal |
| `<`         | Less than |
| `<=`        | Less than or equal |
| `>`         | Greater than |
| `>=`        | Greater than or equal |

Primitive types are compared using signed comparison, except for the following types:

* `false` is less than `true` for `boolean`
* `fixed` and `binary` use unsigned byte-wise comparison
* `string` uses unsigned byte-wise comparison of the UTF-8 representation; it is not the Unicode Collation Algorithm
* `uuid` uses unsigned byte-wise comparison of the UUID bytes

> * `float` and `double` use IEEE 754 total order after normalizing NaN to the canonical NaN (sign bit 0, exponent bits all 1, mantissa msb 1 followed by all 0)

Reviewer 2: I thought I added a comment here before but I was wondering why we want the cannonical nan here when we don't use it this way in the Table Format spec (-Nan < Nan )

Author: My original answer is here: https://github.com/apache/iceberg/pull/16652#discussion_r3358984449

I'd probably rephrase that to make the logic a bit more clear though. First, table stats don't allow NaN values for lower or upper bounds. NaN values are counted and the bounds reflect all non-NaN values. That is because NaN values can easily destroy the effectiveness of stats, since any NaN value is greater than (or -NaN is less than) all other floating point values. I think that the rules here for expressions are as consistent with this as we can be.

For expressions, we want to handle NaN like null: we don't want wacky comparison behavior and we want everyone to use `IS NaN` and `IS NOT NaN` to explicitly specify how to handle NaN values.

What do I mean by "wacky comparison behavior"? In the IEEE 754 non-total-order case, `NaN = NaN` is false and comparisons of values with NaN are always false. You get the bad behavior where the order of comparison matters (`a < NaN` is false and `NaN < a` is false, but a is not NaN) and so NaN values can show up in either upper or lower bounds (and sometimes both!). IEEE 754 total order isn't much better: `NaN = NaN` is true if they have the same bits, but `NaN < NaN` can also be true.

These quirks are not obvious, which is why we typically want people to use the `IS NaN` and `IS NOT NaN` forms. However, just like we have to handle null values produced by functions, we also have to handle NaN values when they show up as the results of functions (like `divide(a, b)` for a=0.0 and b=0.0).

Because we can't completely eliminate NaN comparisons, we have to define how they work. I think the most predictable and understandable way is using the two rules just below: `NaN = NaN` is true for any two NaN values, and all non-NaN values are less than NaN.

One additional choice I made was not to handle -NaN and NaN separately. I think -NaN isn't very useful so I chose the simpler option to normalize _all_ NaN values.

Author: After thinking about this more and talking offline with @RussellSpitzer, we decided to update the wording so that NaN behavior is the same as null handling. This consistency makes both nominally incomparable cases behave the same way.

This could cause situations like those mentioned above where aggregations with NaN values behave strangely based on the order of operands. I think consistency is more important and that NaN handling should always be explicit using `IS NaN`, just like `IS NULL` is the correct way to handle null values. In addition, the use cases contemplated by this spec (for instance `CHECK` constraints) are unlikely to hit these cases.

This specification covers the structure of Iceberg expressions and includes appendices that specify serialization as JSON and a set of portable functions defined by Iceberg specifications.

[udf-spec]: https://iceberg.apache.org/udf-spec

## Structure

Iceberg expressions have two types:

* **Value expressions** represent data values and transformations of values (function calls) that produce any Iceberg type
* **Predicates** represent comparisons of value expressions as well as combinations of predicates with boolean logic (and, or, not)

### Value expressions

A value expression is an expression that produces a typed value.

Value expressions can be one of three types: a constant value, a field reference, or a function applied to zero or more value expressions.

> #### Constant values

Reviewer 3: very minor, we may just want to consistently use one term throughout this. I think everyone understands constants and literals are the same but since grammars tend to be pedantic, may just be better off sticking to one term.

Author: My rationale is that "constant" is more clear for the purpose of the spec, but "literal" is used in expression libraries. They are basically synonyms so I thought that it was fine to use both. It is okay for implementations to continue using "literal" for constants.

I like using "constant" because the reader knows what it is and how it interacts with functions, without bringing in the questions like "what literal forms are allowed?"

A constant or literal is the simplest type of value expression that represents a specific typed value.

#### Field reference

A field reference represents the value of a specific field in a row. When an expression is evaluated on a row, it returns the value of the field.

Field references may be named references (unbound) or ID references (bound). ID references identify a field by field ID from a schema. Named references identify a field by name that must be resolved to an ID (bound to a schema) to access the field.

> ID references are used for stored expressions, where the identity of the column is determined when the stored expression is created. For example, column constraints are tied to field IDs so that renaming a column does not invalidate the reference in its stored constraint.

Reviewer 5: [orthogonal] : it would be nice to have IDReference in the rest yaml too

Author: Yes, we will add this when we update the yaml.

Named references are used when the identity of the column is determined when the expression is evaluated. For example, query filters are resolved each time a query runs so server-side planning uses unbound named references.

The context in which an expression is used determines the type of references that are valid. Iceberg specifications should document whether ID references, named references, or both are allowed.

> #### Apply function

Reviewer 5: optional: should we add a restriction that if engines is unable resolve the function, lets say the function deleted or no longer exists in catalog, MUST fail execution ?

Author: I think that this is dependent on the context and should be specified. For example, we state in the table spec that unknown partition transforms should be ignored, rather than causing reads to fail. That was added specifically to create the ability to add new partition transforms without breaking forward-compatibility. Adding a blanket rule here would conflict with those decisions to I think these cases should be decided by the context where an expression is used or stored.

> An apply expression represents the result of a function applied to (or called on) zero or more values produced by child value expressions or predicates.

Reviewer 3: Minor: Do we need the "or called on"? Doesn't applied to already cover that or is there a distinction we're trying to make?

Author: These are synonyms but I just want to be clear. This also makes it clear why the name is an "apply" expression rather than a "function" expression.

> Functions are referenced using a catalog and a function identifier (list of strings).

Reviewer 3: Minor: 
function identifier which is a list of strings.

Reviewer 3: Or given the below section, we might not even need to have this

Author: Updated by removing it.

> * The function identifier consists of 0 or more namespace names followed by the function name. At least one part, the function name, is required.

Reviewer 4: This prose rule ("at least one part, the function name, is required") is not enforced by the grammar in Appendix B, which uses `[ NAME* ]` for all three structured `FUNC_REF` forms:

```
FUNC_REF: NAME
    | [ NAME* ]
    | { "identifier": [ NAME* ] }
    | { "catalog": NAME, "identifier": [ NAME* ] }
```

Use `[ NAME+ ]` (one-or-more) so the grammar matches the prose:

```
FUNC_REF: NAME
    | [ NAME+ ]
    | { "identifier": [ NAME+ ] }
    | { "catalog": NAME, "identifier": [ NAME+ ] }
```

Author: I don't think it's a problem for the grammar to be wider than what we actually allow, but I think we can tighten it up if this is important.

> * Catalog is optional and is assumed to be the catalog in which the referencing object is stored if it is not present or is null

Reviewer 3: Would it be better to require ommitting the catalog in those cases instead of having readers also to consider the explicit null case?

Author: I don't think so. I think we don't want to need to later define how to handle a `null` catalog. I would rather just specify that they are the same thing.

The catalog name identifies the catalog where the function definition can be loaded or is a reserved name that identifies a set of functions. As in the view and UDF specs, catalog names represent connection configurations that may differ across environments. Omitting catalog names is recommended to avoid depending on consistent environments. For example, if a table has a CHECK constraint that references a UDF without a catalog name (missing or null), the UDF should be loaded from the table's catalog.

The reserved names used to identify sets are:

* `sql_functions` is used for functions defined by the SQL standard
* `iceberg_functions` is used for functions defined in this specification

> Engines may document and use a catalog name to identify their built-in functions that are not part of the SQL spec, like `spark_builtin_functions.to_utc_timestamp`.

Reviewer 7: So we have the risk that a non-interchangeable engine-specific expression used by a Iceberg table would ruin its interoperability?

Author: Expressions are not engine-specific, but functions may be.

That said, there are cases where users want to store engine-specific functions and I think we have to support it. We are planning to implement this using a property to enable non-portable functions that are outside of the SQL and Iceberg sets. And even then, we don't expect all implementations to be able to apply all of the functions from the SQL spec. The principle we're applying here is to nudge toward interoperability and compatibility, but not force it and limit the effectiveness of the features.

Function references are unambiguous and are not interpreted using session context. Producers are responsible for resolving catalog, namespace, and name if the session is relevant. For example, if a SQL engine uses its current catalog and namespace to find a function, the resolved catalog and namespace must be used to produce an unambiguous function reference.

#### Value expression types

The type produced by a value expression may change. For example, an ID reference may produce a widened type after the underlying column's type is promoted.

A value expression's result type is determined when it is bound to a specific input schema.

Function calls may produce different types when function definitions change, and type changes may change the definition that is resolved for a function name. For example, if the input field passed to `identity(int) -> int` is promoted from `int` to `long`, the resolved `identity` function can change to `identity(long) -> long` if it is defined.

> If types are incompatible at runtime, implementations binding or evaluating expressions may apply type promotion to align types for predicates and to resolve functions. Implementations may choose when to promote values to accommodate engines that differ in casting behavior. However, implementations must fail rather than insert unsafe casts.

Reviewer 7: Is `type promotion` here identical to iceberg type promotion rule? I assume no.

Author: It is not. Type promotion is delegated to implementations: "Implementations may choose when to promote values to accommodate engines that differ in casting behavior"

#### Tests

Tests are predicates that test a single value expression, optionally using a constant or set of constants. Constants must all have the same type and must be non-null and non-NaN. Tests are:

| Test                    | Allowed types | Constant type | Description |
|-------------------------|---------------|---------------|-------------|
| `IS NULL`               | any           |               | true iff the value is null |
| `IS NOT NULL`           | any           |               | true iff the value is not null |
| `IS NaN`                | float, double |               | true iff the value is an IEEE 754 NaN |
| `IS NOT NaN`            | float, double |               | true iff the value is not an IEEE 754 NaN |
| `STARTS WITH const`     | string        | string        | true iff the constant is a prefix of the value |
| `NOT STARTS WITH const` | string        | string        | true iff the constant is not a prefix of the value |
| `IN (constant set)`     | any primitive | same as value | true iff the value is equal to any constant |

> | `NOT IN (constant set)` | any primitive | same as value | true iff the value is not equal to all constants |

Reviewer 8: nit: not equal to any constant.

Author: Fixed.

| Comparison     | Result  |
|----------------|---------|
| `null = null`  | `true`  |
| `34 = null`    | `false` |
| `null != null` | `false` |
| `34 != null`   | `true`  |
| `null < null`  | `false` |
| `null <= null` | `true`  |
| `34 < null`    | `false` |

Value expressions that are the direct child of a comparison must not be either a null or NaN constant. However, comparisons must handle null and NaN values that are the result of evaluating a value expression. For example, `x = get_item(map, "key")` is valid although `get_item` may return a null value, but `x = null` must be rejected because `x IS NULL` is the correct unambiguous predicate. Similarly, `multiply(a, b)` may produce NaN for `a=0.0` and `b=Infinity` and is valid, but `x = NaN` must be rejected because `x IS NaN` is the correct test.

Primitive types are compared using signed comparison, except for the following types:

* `false` is less than `true` for `boolean`
* `fixed` and `binary` use unsigned byte-wise comparison
* `string` uses unsigned byte-wise comparison of the UTF-8 representation; it is not the Unicode Collation Algorithm
* `uuid` uses unsigned byte-wise comparison of the UUID bytes

> * `decimal` uses signed comparison independent of scale; this is equivalent to comparison of unscaled values because type alignment produces values with the same scale

Reviewer 7: Not a native speaker but the phrase "independent of scale" seems to contradict "with the same scale". This might confuse developers.

Author: This is saying that type alignment produces the same scale, so they are equivalent.

* `float` and `double` use IEEE 754 order for all non-NaN values; see below for NaN comparison rules

For floating point values, comparison with NaN behaves similarly to comparison of values with null. NaN should be specifically handled using `IS NaN` and `IS NOT NaN` tests. However, when value expressions produce a NaN value, the following rules must be applied:

* `a = b` is true if both are NaN, or both are non-NaN and equal; false otherwise
* `a != b` is the boolean negation of `a = b`

> * `a < b` and `a > b` are false when either operand is NaN; otherwise the IEEE 754 total order is used

Reviewer 4: I think this should say `IEEE 754 order`, not `IEEE 754 total order`. The two terms diverge on ±0 even with the NaN guard on this line:

- Default comparison ([IEEE 754][ieee] §5.11): `−0 == +0`.
- `totalOrder` predicate (§5.10): `−0 < +0`.

Spark SQL, Trino, Postgres, and Java's `<` operator follow default (`−0 == +0`); Java's `Double.compare` / `Arrays.sort(double[])` follow totalOrder (`−0 < +0`). 

The spec picked the rule in L178. We should match it here.

```suggestion
* `a < b` and `a > b` are false when either operand is NaN; otherwise the IEEE 754 order is used
```

[ieee]: https://en.wikipedia.org/wiki/IEEE_754

Reviewer 7: It seems that `IEEE 754 total order` is much clearer and well defined. One can simply write `x = -0 || x = +0` to mimic semantics of `IEEE 754 order` but cannot tell them apart if `IEEE 754 order` is used.

Author: Yes, this was an accident. I intended to use IEEE 754 order because comparison considers -0.0 equal to +0.0.

* `a <= b` is `(a = b) OR (a < b)`; `a >= b` is `(a = b) OR (a > b)`; both are true when both operands are NaN and false when only one operand is NaN

#### Boolean logic

Predicates must use 2-valued boolean logic. Evaluation of all predicates must produce `true` or `false`.

> Engines that implement SQL 3-valued boolean logic must add `IS NULL` and `NOT NULL` to produce the 2-valued equivalent. This avoids bugs in engines and languages that do not natively implement 3-valued logic. For example, the SQL predicate `x < 10` should be passed as `x < 10 AND x IS NOT NULL` for a SQL `WHERE` condition (or `x < 10`; see null-safe comparisons below). For a `CHECK` constraint, the expression is passed as `x < 10 OR x IS NULL`. This ensures that implementations will make the correct determination, rather than depending on context to interpret a null result (`WHERE` vs `CHECK`).

Reviewer 6: nit: "must add `IS NULL` and `NOT NULL`" -> "must add `IS NULL` and `IS NOT NULL`"

Author: Fixed.

The deprecated expressions were passed in 3 places:

* As `filter` passed to server-side scan planning
* As `filter` passed to the service in `ScanReport`
* As `residual` passed to the client with a scan task

Both server-side scan planning and the report endpoint should continue to accept filters from older clients by parsing term-based expressions (see [Appendix B: JSON serialization](#backward-compatibility)).

Residuals passed from services back to clients that do not use the new syntax would cause clients to fail. Services are allowed to omit the residual so that it is calculated on the client side (intended to avoid duplicating large IN filters). For compatibility, REST services should omit residuals from tasks, but may include them if the service detects support for newer predicates (for example, via client version).

## Appendix A: Iceberg functions

This section defines the functions in the `iceberg_functions` reserved catalog name.

* `if_else(condition: predicate, when_true: T, when_false: T) -> T`: returns the value of `when_true` when `condition` is true and `when_false` otherwise

> ### Partition transforms

Reviewer 3: Does having partition transforms here as well as in the main spec put us in a position where any changes to these specs have to make 2 updates? Or are we planning on removing it from the main page as well? 

Also on that note, I think we're missing multi-argument transforms here.

Author: These are the corresponding functions. I think we will want to find a good way to keep these in sync, but I didn't want to remove things from the table spec as part of this change. I do think it would be reasonable to move some of those definitions here.

For multi-arg transforms, that is purely a table spec concept. These are functions, some of which take two args, but these do not define `source-id` or `source-ids`. I think we will want to define the functions here and then map the transform serialization into arguments to these functions.

Iceberg partition transforms are also defined as functions (other than `void`).

All partition transforms produce `null` for a `null` input value.

| Function name     | Description                                                  | Source types                                                         | Result type |
|-------------------|--------------------------------------------------------------|----------------------------------------------------------------------|-------------|

> | `identity(value)` | Source value, unmodified                                     | Any primitive except for `geometry`, `geography`, and `variant`      | Source type |

Reviewer 7: This is related to https://github.com/apache/iceberg/pull/16836. Should we remove `variant` because it is not a primitive type? (cc @zhjwpku)

Reviewer 9: Yeah, I also noticed that this section uses `Function name`, while `format/spec.md` uses `Transform name`. Also, here the function name includes the parameter signature, whereas `format/spec.md` only uses the transform name without parameters. Not a strong opinion but should we keep them consistent?

Author: Variant is not allowed in an identity transform. This is not changing anything from the table spec, it is just adding a function definition that corresponds to the existing transform.

| `year(value)`     | Extract a date or timestamp year, as years from 1970         | `date`, `timestamp`, `timestamptz`, `timestamp_ns`, `timestamptz_ns` | `int`       |
| `month(value)`    | Extract a date or timestamp month, as months from 1970-01-01 | `date`, `timestamp`, `timestamptz`, `timestamp_ns`, `timestamptz_ns` | `int`       |

> | `day(value)`      | Extract a date or timestamp day, as days from 1970-01-01     | `date`, `timestamp`, `timestamptz`, `timestamp_ns`, `timestamptz_ns` | `date`      |

Reviewer 8: Should the type returned here by int rather than date since it's number of days from 1971-01-01?

Reviewer 9: This has been discussed recently, you may see #16414

| Parameterized function name | Description                                                           | Source types                                                                                 | Result type |
|-----------------------------|-----------------------------------------------------------------------|----------------------------------------------------------------------------------------------|-------------|
| `bucket(N, value)`          | Hash of value, mod `N` (see [table spec details][bucket-ref])         | Any primitive except for `geometry`, `geography`, `variant`, `boolean`, `float`, or `double` | `int`       |
| `truncate(W, value)`        | Value truncated to width `W` (see [table spec details][truncate-ref]) | `int`, `long`, `decimal`, `string`, `binary`                                                 | Source type |

[bucket-ref]: spec/#bucket-transform-details
[truncate-ref]: spec/#truncate-transform-details

## Appendix B: JSON serialization

Iceberg expressions are serialized as JSON objects in table, view, and UDF metadata, and in the REST protocol for catalogs.

### Value expressions

```
EXPR: LITERAL | REFERENCE | APPLY

LITERAL: VALUE

>     | { "type": "literal", "value": VALUE }

Reviewer 3: In the new grammar do we still want to allow untyped literals? I understand we may need to do it for compatibility reasons but I was under the impression here we'd deprecate producing the untyped one (so DEPRECATED_LITERAL, and then the new one with `data-type`. I'm mostly a bit concerned about avoiding interpretability issues of literals and producing an explicit Iceberg type doesn't seem like a significant burden? Unless there are cases where it's not reasonably known?

Author: I think it is fine to do this, since the values are still going to be bound to align types at runtime.

    | { "type": "literal", "value": VALUE, "data-type": DATA_TYPE }

> LITERALS: [ LITERAL* ]

Reviewer 7: Should we design a simplified form for `LITERALS`? It seems unnecessary to have repeated `"type"` and `"data-type"` provided that they share the same type. This can help save some bytes for a `SET_OP` with many literals.

Author: I think that's reasonable.

UNBOUND_REF: { "type": "reference", "name": NAME }

APPLY: { "type": "apply", "function": FUNC_REF, "arguments": [ FUNC_ARG* ] }
FUNC_ARG: EXPR | PREDICATE
FUNC_REF: NAME
    | [ NAME* ]
    | { "identifier": [ NAME* ] }
    | { "catalog": NAME, "identifier": [ NAME* ] }

ID: integer
NAME: string

VALUE: single value JSON from the table spec
DATA_TYPE: Iceberg type from the spec
```

If a function reference is a string, that string is the one-part identifier and the catalog is missing/null.
If a function reference is a list of strings, it is the function identifier and the catalog is missing/null.

### Predicates

```
PREDICATE: true | false
    | { "type": "not", "child": PREDICATE }
    | { "type": BINARY_OP, "left": PREDICATE, "right": PREDICATE }
    | { "type": UNARY_OP, "child": EXPR }
    | { "type": CMP_OP, "left": EXPR, "right": EXPR }
    | { "type": SET_OP, "child": EXPR, "values": LITERALS }
    | DEPRECATED_PREDICATE

BINARY_OP: "and" | "or"
UNARY_OP: "is-null" | "not-null" | "is-nan" | "not-nan"
CMP_OP: "lt" | "lt-eq" | "gt" | "gt-eq" | "eq" | "not-eq"
      | "starts-with" | "not-starts-with"
SET_OP: "in" | "not-in"
```

### Backward compatibility

```

> DEPRECATED_PREDICATE:

Reviewer 10: how about `{ "type": "true" }` and `{ "type": "false" }`?

Author: Is this currently allowed?

Reviewer 10: https://github.com/apache/iceberg/blob/a2fb64bff8ea2c299d56f81d666778bf4365fdac/open-api/rest-catalog-open-api.yaml#L2562-L2578

Reviewer 10: I created a folllow-up PR https://github.com/apache/iceberg/pull/17120. Please help review.
